src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -532,43 +526,11 @@ public override void SetAsyncCallbacks(SNIAsyncCallback receiveCallback, SNIAsyn
         /// <returns>SNI error code</returns>
         public override uint SendAsync(SNIPacket packet, SNIAsyncCallback callback = null)
         {
-            SNIPacket newPacket = packet;
-
-            _writeTaskFactory.StartNew(() =>
+            SNIAsyncCallback cb = callback ?? _sendCallback;
+            lock(this)
             {
-                try
-                {
-                    lock (this)
-                    {
-                        packet.WriteToStream(_stream);
-                    }
-                }
-                catch (Exception e)
-                {
-                    SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, SNICommon.InternalExceptionError, e);
-
-                    if (callback != null)
-                    {
-                        callback(packet, TdsEnums.SNI_ERROR);
-                    }
-                    else
-                    {
-                        _sendCallback(packet, TdsEnums.SNI_ERROR);
-                    }
-
-                    return;
-                }
-
-                if (callback != null)
-                {
-                    callback(packet, TdsEnums.SNI_SUCCESS);
-                }
-                else
-                {
-                    _sendCallback(packet, TdsEnums.SNI_SUCCESS);
-                }
-            });
-
+                packet.WriteToStreamAsync(_stream, cb, SNIProviders.TCP_PROV);
+            }
```

Comment:
Same question I had before... this is now potentially invoking `cb` while holding the lock... is that ok?

Reply:
I could remove the locks from async reads and writes. As a result none of the operations are now holding a lock. I had another change for removal of locks during async reads and writes being tested. I clubbed the change with this PR. Please see comment at https://github.com/dotnet/corefx/pull/27743#issuecomment-370609694